lib/stdlib/src/rand.erl

```diff
@@ -1412,6 +1424,83 @@ dummy_seed({A1, A2, A3}) ->
     {Z3, _} = splitmix64_next(A3 bxor X2),
     ?MASK(58, Z3).
 
+
+%% =====================================================================
+%% mcg35 PRNG: Multiplicative Linear Congruential Generator
+%%
+%% Parameters by Pierre L'Ecuyer from the paper:
+%%   TABLES OF LINEAR CONGRUENTIAL GENERATORS
+%% OF DIFFERENT SIZES AND GOOD LATTICE STRUCTURE
+%%
+%% X1 = (A * X0) rem M
+%%
+%% A = 185852
+%% M = 2^B - D
+%%     B = 35, D = 31
+%%
+%% X cannot, and never will be 0.
+%% Take X1 as output value with range 1..(M-1).
+%%
+%% Use this generator for speed, not for quality.
+%% The calculation should avoid bignum operations,
+%% so A * X0 should not become a bignum.
+%% M and A has been selected accordingly.
+%% The period is M - 1 since 0 cannot occur.
+%%
+%% =====================================================================
+-define(MCG35_A, (185852)).
+-define(MCG35_B, (35)).
+-define(MCG35_D, (31)).
+-define(MCG35_M, (?BIT(?MCG35_B) - ?MCG35_D)).
+
+-type mcg35_state() :: 1..(?MCG35_M-1).
+
+-spec mcg35(X0 :: mcg35_state()) -> X1 :: mcg35_state().
+mcg35(X0) ->
+    X = ?MCG35_A * X0,
+    %% rem M = rem (2^B - D), optimization to avoid 'rem'
+    X1 = ?MASK(?MCG35_B, X) + ?MCG35_D*(X bsr ?MCG35_B),
+    if
+        X1 < ?MCG35_M -> X1;
+        true          -> X1 - ?MCG35_M
+    end.
+
+
+%% =====================================================================
+%% lcg35 PRNG: Multiplicative Linear Congruential Generator
+%%
+%% Parameters by Pierre L'Ecuyer from the paper:
+%%   TABLES OF LINEAR CONGRUENTIAL GENERATORS
+%% OF DIFFERENT SIZES AND GOOD LATTICE STRUCTURE
+%%
+%% X1 = (A * X0 + C) rem M
+%%
+%% M = 2^35, A = 15319397, C = 1
+%%
+%% Choosing C = 15366142135, which is an odd value
+%% about 2^35 / sqrt(5) gives a perhaps nicer value in
+%% the sequence after 0 (-> C).
+%%
+%% Take X1 as output value with range 1..(M-1).
+%%
+%% Use this generator for speed, not for quality.
+%% The calculation should avoid bignum operations,
+%% so A * X0 + C should not become a bignum.
+%% M, A and C has been selected accordingly.
+%% The period is M.
+%%
+%% =====================================================================
+-define(LCG35_A, (15319397)).
+-define(LCG35_C, (15366142135)). % (1 bsl 35) / sqrt(5), odd
+-define(LCG35_B, 35). % Number of bits
+
+-type lcg35_state() :: 0..?MASK(?LCG35_B).
+
+-spec lcg35(X0 :: lcg35_state()) -> X1 :: lcg35_state().
+lcg35(X0) ->
+    ?MASK(?LCG35_B, ?LCG35_A * X0 + ?LCG35_C).
```

Comment:
Here is an idea for possible improvement:

```suggestion
    ?MASK(?LCG35_B, ?LCG35_A * ?MASK(?LCG35_B, X0) + ?LCG35_C).
```

With this change, an unnecessary masking operation will be performed, but because of it, the type tests, overflow tests, and fallback code for bignum arithmetic will be omitted for the remaining three operations, resulting in smaller code with only a single branch instruction.

I don't know whether this will be faster. Perhaps you could re-run your benchmarks and find out.

If you want to inspect the native code, start the runtime system like this:

```
erl +JDdump true
```

The native code will be dumped to the `rand.asm` file.

Reply:
I tried and the benchmark maybe shows some improvement, it is hard to measure in the 2..4 nanoseconds range, but the generated native code gets significantly shorter, as you anticipated.  Using the same trick on the `mcg35` generator seems to give an even better improvement probably since there is a bit more code after the masking.

I also tried a better looking change i.e to add a guard `when is_integer(X0), 0 =< X0, X0 =< ?MASK(?LCG35_B)`, but apparently the JIT did not use that information to optimize the function body :-( Furthermore the 3-part range test itself needs much more code than a `band` operation which was a setback...

I dream of an integer_range guard test in the VM that the compiler could use in these cases, and that such a guard would be used by the JIT for function body optimization!

For both generators such a guard on the input would be user friendly anyway, especially for the `mcg35` generator where the allowed range is 1 =< X < 2^35 - 31 where masking with 2^35 - 1 is goofy.

Reply:
> I also tried a better looking change i.e to add a guard when `is_integer(X0), 0 =< X0, X0 =< ?MASK(?LCG35_B)`, but apparently the JIT did not use that information to optimize the function body :-( Furthermore the 3-part range test itself needs much more code than a `band` operation which was a setback...

No promises, but we hope to address both of those issues in OTP 26.